### DISCUSSION TITLE
How to map PDU with failed status from submit_sm_resp back to request?

### DISCUSSION THREAD

**Discussion by mslatour** [python-smpplib/python-smpplib#144] [Q&A] [answered]
We've been using our own fork of this library for a few years now and I'm trying to see if we can move back to using the main repo again. There is one blocker that I cannot yet figure out.

When a submit_sm_resp contains a SMPP error, this could indicate the status of the preceding submit_sm request we sent to the SMSC. However looking at your current implementation, it seems that a PDUError would be raised upon receiving such a 'failed' response that only contains the command name and SMPP status. Apart from potentially killing a listening loop, I don't see how this allows us to know what request this was a response to? And so how would we know that a certain SMS failed to be posted, given that it is async.

Not sure if it is uncommon for the SMPP status field to be used to indicate the submission failure, but our telco's SMSC certainly does this. Also looking at the different status values, this doesn't seem that weird to me.

1) How would you handle this?
2) Is there a good reason why the exception does not just include the entire PDU object or otherwise at least the sequence number? So when the exception bubbles up there can be logic processing the sending failure.

Help is much appreciated.

**Comment by mslatour** [comment]
Actually looking at it again. I don't see why this would be a PDUError at all, as that exception is documented to mean "Error processing PDU". And there is no problem processing the PDU, the packet just indicates that the indented operation failed. 

For now I'll just extend Client with our own wrapper class that implements different behaviour around this and keep moving forward. Probably just passing through the PDU to the handlers and have them deal with if the PDU indicates a failure or not. And only throwing PDUError if the PDU itself is not valid or couldn't be fully read.

I'm still interested in your insights into this as I'm sure you have more context on this than I do.  If you think my approach is what you would also want, then I'm happy to submit a PR but otherwise I'll just do my override for our own code.

Thanks for bringing the SMPP library this far already, it saves us a lot of work!

**Comment by eigenein** [comment]
Hi,

Thanks for pointing this out!

Yes, indeed, the package contains a lot of legacy code that isn't touched unless there was a reason to. Perhaps, no one has just tried to trace an error back to request, but that's a useful functionality.

From what I can see, it's indeed not really correct to use `PDUError` for both command error and PDU format error. I agree these are two completely different kinds of errors.

Could you show us your patch? I believe we could incorporate things into `smpplib` so that you wouldn't need to maintain the patched version.

**Comment by mslatour** [comment]
Hi @eigenein,

Sorry I wasn't monitoring this thread actively anymore.

We don't maintain it as a forked repo anymore, but this is the patch extracted from our code. Basically just ignoring if the PDU indicated an error (but was successfully parsed) and leaving that up to the caller higher up to handle. We've kept the function signature the same but the `ignore_error_codes` field also became obsolete. Other than that the code is exactly the same I think (since we last pulled smpplib)

```
class WrapperSMPPClient(smpplib.client.Client):
    """Wrapper client around smpplib.client.Client to override some behaviours."""

    def read_once(self, _ignore_error_codes: Optional[List[str]] = None) -> None:
        """Read a PDU and act.

        We've overrided this method to not raise an PDUError when the PDU indicates
        a failed status. Instead we just want those to be provided to the handlers.
        """
        try:
            pdu = self.read_pdu()
        except socket.timeout:
            logger.debug("Socket timeout, listening again")
            pdu = smpplib.smpp.make_pdu("enquire_link", client=self)
            self.send_pdu(pdu)
            return

        if pdu.command == "unbind":  # unbind_res
            logger.info("Unbind command received")
            return
        elif pdu.command == "submit_sm_resp":
            self.message_sent_handler(pdu=pdu)
        elif pdu.command == "deliver_sm":
            self._message_received(pdu)
        elif pdu.command == "enquire_link":
            self._enquire_link_received()
        elif pdu.command == "enquire_link_resp":
            pass
        elif pdu.command == "alert_notification":
            self._alert_notification(pdu)
        else:
            logger.warning('Unhandled SMPP command "%s"', pdu.command)
```

Does that help?

**Comment by davidshepherd7** [comment]
@eigenein Hi, I've been looking at this and I have two ideas for a solution without making a breaking change:

1. Add another boolean option to the client, something like `raise_on_error_pdus` and default to current behaviour.

2. Add two new Exception classes which inherit from PDUError: PDUParsingError and PDUErrorResponse (and include the PDU itself as a member of the error response exception).

I think if I was implementing this from scratch I would go for the behaviour given by 1 (most callback-driven APIs that I've used call the callback with an error flag on most error cases). But I think 2 interacts much better with the current design: it doesn't change the current "contract" where handlers are only called with successful PDUs (which would be a big potential source of bugs for users) and it interacts sensibly with the existing `ignore_error_codes` argument.

What do you think?

**Comment by eigenein** [comment]
Hi David,

I'm more in favour of option 2 because of the arguments you listed. And the we'd be able to attach original PDU instead of blindly ignoring it.

@podshumok Do you have an opinion on that too?

**Comment by code-of-kpp** [comment]
First of all, subclassing `Client` is OK and we have to keep that in mind when we do any changes.

Having that said, I think the best way would be to provide an error handler similar to other handlers we have. It should raise by default if code is not in `ignore_error_codes` param which should be gradually be excluded from `read_pdu*` signature

**Comment by eigenein** [comment]
I don't really get what you mean, is that 1, 2 or a different option?

**Comment by code-of-kpp** [comment, marked as answer]
:) Another one.

now we have:
```python
            if pdu.is_error():
                raise exceptions.PDUError('({}) {}: {}'.format(
                    pdu.status,
                    pdu.command,
                    consts.DESCRIPTIONS.get(pdu.status, 'Unknown status')),
                    int(pdu.status),
                )
```

I think it should be:
```python
            if pdu.is_error():
                self.error_pdu_handler(pdu, *someotheroptionsmaybe)
```

where `client.error_pdu_handler = default_error_pdu_handler` and
```python
            def default_error_pdu_handler(pdu, *args):
                raise exceptions.PDUError('({}) {}: {}'.format(
                    pdu.status,
                    pdu.command,
                    consts.DESCRIPTIONS.get(pdu.status, 'Unknown status')),
                    int(pdu.status),
                )
```

**Comment by davidshepherd7** [comment]
@podshumok Yeah I like this design better, I'll have a go at implementing it now.

I think the resulting API is almost as good as option 1, but it's also much less likely to cause backwards incompatibility problems.
